### PR TITLE
Project Auth - Axel Hammarbäck & Vanessa Sue Smith

### PR DESCRIPTION
**Pull request by VanessaSue27**
Project Auth submission 🔐: Axel Hammarbäck and Vanessa Sue Smith

Deployed project:
Frontend: https://project-auth-vane-axel-frontend.netlify.app/
Backend: https://project-auth-vane-axel.herokuapp.com/

This was a very interesting project to work on, seems security is a topic to be extremely aware of, yet very wide, but good to have started with a basic project to understand the start of it 💪

We took some time to test our app a lot, to make sure all the validation was kicking in (both in the Backend and Frontend), mainly to prevent any weird username and password combinations from being saved to our database.

It was cool going back to Redux and refreshing those concepts, and once more a joy connecting the Backend to the Frontend and see that final product run as intended 🤩

// Vanessa and Axel

### CODE REVIEW

File: frontend/src/pages/SecretPage.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  // and take us back to the sign up page
+  const handleClick = () => {
+    dispatch(logout());
+    setPage('signup');
+  };
+
+  return (
+    <>
+      <h1>Shhhh! Secret Page...</h1>
+      <h3>{`${secretMessage}`}</h3>
```
Comment (ClaireCaudwell) [inline frontend/src/pages/SecretPage.js:38]: Just wondering why you've written a template literal here and not just curly brackets? 😊

File: frontend/src/pages/SignUpPage.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        return res.json();
+      })
+      .then((json) => {
+        dispatch(user.actions.setUserId({ userId: json.userId }));
+        dispatch(user.actions.setAccessToken({ accessToken: json.accessToken }));
+      })
+      .catch((error) => {
+        dispatch(user.actions.setErrorMessage({ errorMessage: error.toString() }));
+      });
+  };
```
Comment (ClaireCaudwell) [inline frontend/src/pages/SignUpPage.js:53]: Just wondering why this fetch is done here and not in the redux store like the others? 😊

File: frontend/src/index.css [hunk 3]
```diff
@@ -1,13 +1,160 @@
...
+}
+.button-primary:hover {
+  background-color: #7e3ff2;
+}
+.button-primary:active {
+  background-color: #5300e8;
+}
+.button-primary:focus {
+  box-shadow: inset 0 0 0 3px #6002ee, inset 0 0 0 5px #fff;
+}
```
Comment (ClaireCaudwell) [inline frontend/src/index.css:90]: Interesting to see how you can do this with the different psudeo classes 👌

**Review by ClaireCaudwell — APPROVED** [review]
Hi Vanessa and Alex 😊

Great job with you login/sign up form. Styling looks amazing and interesting to see how you can toggle between the sign and login components. Also interesting to see that you do the fetch for the GET request to show the secret page straight after doing the POST requests. 

Added comments with regards to error messages, so maybe good to check those out. 

All in all, great job as usual 👍

Anna, Claire and Sabina 😊

File: backend/server.js [hunk 4]
```diff
@@ -1,29 +1,132 @@
...
+  try {
+    const { name, password } = req.body;
+    const user = await User.findOne({ name });
+
+    if (user && bcrypt.compareSync(password, user.password)) {
+      // if the user is found and the password matches, we respond with the user ID and their access token
+      res.status(201).json({ userId: user._id, accessToken: user.accessToken });
+    } else {
+      res.status(404).json({ notFound: true, message: 'Verify username and password is correct' });
+    }
```
[the remark this answers is not in the document]
Comment (KarolCislowski) [reply backend/server.js:123]: You can use `throw` to kick out your flow from `try` and then the `catch` will happened. 

https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Polecenia/throw

File: frontend/src/pages/LogInPage.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+import { user, login } from '../reducers/user';
+
+// This page shows a form in which the user can fill up their username
+// and password, those values will be saved in local storage variables
+// When the "LOG IN" button is clicked, this will dispatch the login thunk
+// which carries the username and password values
+// login will do a fetch to the sessions endpoint, which will validate the user
+// if successful, it will save the user's ID and access token in the global
+// store, giving the user access to the secret page
```
Comment (KarolCislowski) [inline frontend/src/pages/LogInPage.js:12]: I love how you documented your code 🥰

File: frontend/src/reducers/user.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    dispatch(user.actions.setAccessToken({ accessToken: null }));
+    dispatch(user.actions.setErrorMessage({ errorMessage: null }));
+    dispatch(user.actions.setSecretMessage({ secretMessage: null }));
+  };
+};
+
+// Article to remove warning: "Expected an error object to be thrown no-throw-literal"
+// https://eslint.org/docs/rules/no-throw-literal
+// How to access the Error message so it can be set as Error Message: toString()
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/toString
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:91]: This reducer is a piece of art!! 😮  But you need to store the token in the local storage to make your user logged when the page is refreshed.

File: frontend/src/pages/SignUpPage.js (hunk 2 above)
Comment (KarolCislowski) [reply frontend/src/pages/SignUpPage.js:53]: Because this fetch is reaching outside the app to the Server :)

File: frontend/src/index.css [hunk 7]
```diff
@@ -1,13 +1,160 @@
...
+  color: #6002ee;
+}
+
+h4 {
+  margin: 16px 0;
+}
+
+p {
+  margin: 0;
+  color: rgba(0, 0, 0, 0.87);
```
Comment (KarolCislowski) [inline frontend/src/index.css:39]: IMHO this text would look nicer with `text-align: justify;`

File: frontend/src/index.css [hunk 8]
```diff
@@ -1,13 +1,160 @@
+/* Primary Purple: #6002EE */
+/* Helpful link for choosing colors: https://material.io/design/color/the-color-system.html#tools-for-picking-colors */
+@import url("https://fonts.googleapis.com/css2?family=Source+Sans+Pro:wght@300;400;600;700&display=swap");
+
+/* Applied to everything, since things scale unintuitively otherwise */
+* {
+  box-sizing: border-box;
```
Comment (KarolCislowski) [inline frontend/src/index.css:7]: 🥇

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
This is an awesome project! 🦄 Nice idea for the secret page! 🦷 
You are coding like ninjas! 🍶 

Please make so your user staid logged in after refresh!

**Comment by VanessaSue27** [comment]
Hello @KarolCislowski ! 🤗
Thanks a bunch for this code review and for pointing out the localStorage and refresh issue. This was very interesting to find out and learn how to fix.
We have now implemented it and it seems to be working as intended 💪

**Review by KarolCislowski — APPROVED** [review]
And now it is perfect 🦄

**Pull request closed without merging**